File: examples/echo_agent_ts/graph.ts
/**
 * Simple echo agent for TypeScript runner validation -- direct TS mirror
 * of examples/echo_agent/graph.py. Receives a user message and echoes it
 * back. Proves the TypeScript runner works with a real LangGraph.js graph,
 * same as the Python spike proved it for LangGraph.
 */
import { Annotation, StateGraph, START, END } from "@langchain/langgraph";

interface Message {
  role: string;
  content: string;
}

const State = Annotation.Root({
  messages: Annotation<Message[]>({
    reducer: (a, b) => a.concat(b),
    default: () => [],
  }),
});

function echoNode(state: typeof State.State) {
  const lastMessage = state.messages[state.messages.length - 1];
  return { messages: [{ role: "ai", content: lastMessage.content }] };
}

const builder = new StateGraph(State).addNode("echo", echoNode).addEdge(START, "echo").addEdge("echo", END);

export const graph = builder.compile();

File: examples/echo_agent/graph.py
"""Simple echo agent for spike validation (VG-001).

Receives a user message and echoes it back. Uses LangGraph StateGraph
to prove the bridge works with real LangGraph graphs.
"""

from typing import Annotated, TypedDict

from langgraph.graph import StateGraph, START, END


class State(TypedDict):
    messages: Annotated[list[dict], lambda a, b: a + b]


def echo_node(state: State) -> State:
    """Echo the last user message back."""
    last_message = state["messages"][-1]
    return {
        "messages": [{"role": "ai", "content": last_message["content"]}]
    }


# Build the graph
builder = StateGraph(State)
builder.add_node("echo", echo_node)
builder.add_edge(START, "echo")
builder.add_edge("echo", END)

graph = builder.compile()
